flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/active/ActiveResourceManagerTest.java

```diff
@@ -0,0 +1,428 @@
+		new Context() {{
+			final Throwable fatalError = new Throwable("Testing fatal error");
+			runTest(() -> {
+				runInMainThread(() -> getResourceManager().onError(fatalError));
+				final Throwable reportedError = getFatalErrorHandler().getErrorFuture().get(TIMEOUT_SEC, TimeUnit.SECONDS);
+				assertThat(reportedError, is(fatalError));
+			});
+		}};
+	}
+
+	class Context {
+
+		final Configuration flinkConfig = new Configuration();
+		final TestingResourceManagerDriver.Builder driverBuilder = new TestingResourceManagerDriver.Builder();
+		final TestingSlotManagerBuilder slotManagerBuilder = new TestingSlotManagerBuilder();
+
+		private ActiveResourceManager<ResourceID> resourceManager;
+		private TestingFatalErrorHandler fatalErrorHandler;
+
+		ActiveResourceManager<ResourceID> getResourceManager() {
+			return resourceManager;
+		}
+
+		TestingFatalErrorHandler getFatalErrorHandler() {
+			return fatalErrorHandler;
+		}
+
+		void runTest(RunnableWithException testMethod) throws Exception {
+			fatalErrorHandler = new TestingFatalErrorHandler();
+			resourceManager = createAndStartResourceManager(
+					flinkConfig,
+					driverBuilder.build(),
+					slotManagerBuilder.createSlotManager());
+
+			try {
+				testMethod.run();
+			} finally {
+				resourceManager.close();
+			}
+		}
+
+		private ActiveResourceManager<ResourceID> createAndStartResourceManager(
+				Configuration configuration,
+				ResourceManagerDriver<ResourceID> driver,
+				SlotManager slotManager) throws Exception {
+			final TestingRpcService rpcService = new TestingRpcService(configuration);
+			final MockResourceManagerRuntimeServices rmServices = new MockResourceManagerRuntimeServices(rpcService, TIMEOUT_TIME, slotManager);
+
+			final ActiveResourceManager<ResourceID> activeResourceManager = new ActiveResourceManager<>(
+					driver,
+					configuration,
+					rpcService,
+					ResourceID.generate(),
+					rmServices.highAvailabilityServices,
+					rmServices.heartbeatServices,
+					rmServices.slotManager,
+					NoOpResourceManagerPartitionTracker::get,
+					rmServices.jobLeaderIdService,
+					new ClusterInformation("localhost", 1234),
+					fatalErrorHandler,
+					UnregisteredMetricGroups.createUnregisteredResourceManagerMetricGroup());
+
+			activeResourceManager.start();
+			rmServices.grantLeadership();
+
+			return activeResourceManager;
+		}
+
+		public void runInMainThread(Runnable runnable) {
+			resourceManager.handleInMainThread(runnable);
+		}
+
+		public <T> CompletableFuture<T> runInMainThread(Callable<T> callable) {
+			return resourceManager.runInMainThread(callable, TIMEOUT_TIME);
+		}
+
+		CompletableFuture<RegistrationResponse> registerTaskExecutor(ResourceID resourceID) throws Exception {
+			final TaskExecutorGateway taskExecutorGateway = new TestingTaskExecutorGatewayBuilder()
+					.createTestingTaskExecutorGateway();
+			((TestingRpcService) resourceManager.getRpcService()).registerGateway(resourceID.toString(), taskExecutorGateway);
```

Comment:
Having a class wide `TestingRpcService` would avoid this casting here.

Reply:
Please see my other comment.